stock_unique_order_per_location/models/stock_move_line.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import _, models
+from odoo.exceptions import UserError
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def _action_done(self):
+        res = super()._action_done()
+        self._check_same_order_at_destination()
+        return res
+
+    def _check_same_order_at_destination(self, location=None):
+        move_lines = self.filtered(
+            lambda line: line.move_id.picking_type_id.same_next_picking
+        )
+        for line in move_lines:
+            destination_location = location or line.location_dest_id
```

Comment:
Don't loop twice over the move lines and don't make a search query for each move line. You should do this check at picking level.
```
for picking_id, move_lines in odoo.tools.groupby(self, lambda line: line.picking_id):
  if not picking_id.picking_type_id.same_next_picking:
    continue
  move_lines = self.browse().union(*move_lines)
  destination_locations = location or move_lines.location_dest_id
  ...
```

Reply:
done